File: src/main/java/com/matamercer/microblog/models/entities/VerificationToken.java
package com.matamercer.microblog.models.entities;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import javax.persistence.*;
import java.sql.Timestamp;
import java.util.Calendar;
import java.util.Date;

@Entity
@Table(name = "verification_token")
@Getter
@Setter
@NoArgsConstructor
public class VerificationToken extends BaseModel{
    private static final int EXPIRATION = 60 + 24;

    private String token;

    @OneToOne(fetch = FetchType.EAGER)
    @JoinColumn(nullable = false, name = "user_id")
    private User user;

    private Date expiryDate;

    public VerificationToken(final String token) {
        super();

        this.token = token;
        this.expiryDate = calculateExpiryDate(EXPIRATION);
    }

    public VerificationToken(final String token, final User user) {
        super();

        this.token = token;
        this.user = user;
        this.expiryDate = calculateExpiryDate(EXPIRATION);
    }

    private Date calculateExpiryDate(int expiryTimeInMinutes) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(new Timestamp(cal.getTime().getTime()));
        cal.add(Calendar.MINUTE, expiryTimeInMinutes);
        return new Date(cal.getTime().getTime());
    }


    public void updateToken(final String token) {
        this.token = token;
        this.expiryDate = calculateExpiryDate(EXPIRATION);
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((getExpiryDate() == null) ? 0 : getExpiryDate().hashCode());
        result = prime * result + ((getToken() == null) ? 0 : getToken().hashCode());
        result = prime * result + ((getUser() == null) ? 0 : getUser().hashCode());
        return result;
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final VerificationToken other = (VerificationToken) obj;
        if (getExpiryDate() == null) {
            if (other.getExpiryDate() != null) {
                return false;
            }
        } else if (!getExpiryDate().equals(other.getExpiryDate())) {
            return false;
        }
        if (getToken() == null) {
            if (other.getToken() != null) {
                return false;
            }
        } else if (!getToken().equals(other.getToken())) {
            return false;
        }
        if (getUser() == null) {
            if (other.getUser() != null) {
                return false;
            }
        } else if (!getUser().equals(other.getUser())) {
            return false;
        }
        return true;
    }

    @Override
    public String toString() {
        final StringBuilder builder = new StringBuilder();
        builder.append("Token [String=").append(token).append("]").append("[Expires").append(expiryDate).append("]");
        return builder.toString();
    }
}

File: src/main/java/com/matamercer/microblog/models/entities/VerificationToken.kt
package com.matamercer.microblog.models.entities

import com.matamercer.microblog.models.entities.BaseModel
import com.matamercer.microblog.models.entities.VerificationToken
import java.lang.StringBuilder
import java.sql.Timestamp
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "verification_token")
class VerificationToken : BaseModel {
    var token: String

    @OneToOne(fetch = FetchType.EAGER)
    @JoinColumn(nullable = false, name = "user_id")
    var user: User? = null
    var expiryDate: Date

    constructor(token: String) : super() {
        this.token = token
        expiryDate = calculateExpiryDate(EXPIRATION)
    }

    constructor(token: String, user: User?) : super() {
        this.token = token
        this.user = user
        expiryDate = calculateExpiryDate(EXPIRATION)
    }

    private fun calculateExpiryDate(expiryTimeInMinutes: Int): Date {
        val cal = Calendar.getInstance()
        cal.time = Timestamp(cal.time.time)
        cal.add(Calendar.MINUTE, expiryTimeInMinutes)
        return Date(cal.time.time)
    }

    fun updateToken(token: String) {
        this.token = token
        expiryDate = calculateExpiryDate(EXPIRATION)
    }

    override fun hashCode(): Int {
        val prime = 31
        var result = 1
        result = prime * result + if (expiryDate == null) 0 else expiryDate.hashCode()
        result = prime * result + if (token == null) 0 else token.hashCode()
        result = prime * result + if (user == null) 0 else user.hashCode()
        return result
    }

    override fun equals(obj: Any?): Boolean {
        if (this === obj) {
            return true
        }
        if (obj == null) {
            return false
        }
        if (javaClass != obj.javaClass) {
            return false
        }
        val other = obj as VerificationToken
        if (expiryDate == null) {
            if (other.expiryDate != null) {
                return false
            }
        } else if (!expiryDate.equals(other.expiryDate)) {
            return false
        }
        if (token == null) {
            if (other.token != null) {
                return false
            }
        } else if (!token.equals(other.token)) {
            return false
        }
        if (user == null) {
            if (other.user != null) {
                return false
            }
        } else if (user!! != other.user) {
            return false
        }
        return true
    }

    override fun toString(): String {
        val builder = StringBuilder()
        builder.append("Token [String=").append(token).append("]").append("[Expires").append(expiryDate).append("]")
        return builder.toString()
    }

    companion object {
        private const val EXPIRATION = 60 + 24
    }
}